Initial draft

jeanconn opened the pull request.

jeanconn committed:
Initial draft

jeanconn committed:
Set to work 'in place' by default

jeanconn committed:
Change fname to filename and idx to line_number

jeanconn committed:
Comment the marks for the warning line in context

jeanconn committed:
Add more fields to copy over

jeanconn committed:
Make line number actually be line number

jeanconn committed:
Add persistent permanent ids to the checks

jeanconn committed:
Add support for the missing field to the saved fields

jeanconn committed:
Print only the context field as a literal block

jeanconn committed:
Save the warning as a string instead of array and print as literal

jeanconn committed:
PEP8 fixes

jeanconn committed:
Make HTML from the list of checks

taldcroft commented on get_checks.py:
```diff
@@ -0,0 +1,189 @@
+    warn_text = []
+    for fline in ftext[idx:]:
+        warn_text.append(fline.strip())
+        if re.search(";", fline):
```
`if ';' in fline`

taldcroft commented on get_checks.py:
```diff
@@ -0,0 +1,189 @@
+    :return: dictionary of the warning
+    """
+    warn_text = []
+    for fline in ftext[idx:]:
```
Use consistent variable names in loops: `for line in lines[idx:]`.  Use this name pretty every time you have a list of text lines.   Always make list objects be named with a plural and then iterate over the singular (unless in the rare case there is some strong reason otherwise).  It is irrelevant here that `lines` (aka `ftext`) came from a file.

taldcroft commented on get_checks.py:
```diff
@@ -0,0 +1,189 @@
+
+def get_single_warn(fname, ftext, idx, context_range=[-15, 10]):
+    """
+    Extra the context and text of a warning
```
`Extract`

taldcroft commented on get_checks.py:
```diff
@@ -0,0 +1,189 @@
+        if re.search(";", fline):
+            break
+    min_line = max(idx + context_range[0], 0)
+    max_line = min(idx + context_range[1], len(ftext) - 1)
```
I don't think this is needed.  I mean clipping it.  In doing this there is always the chance of making an off-by-one error, and I think you have done that because `ftext[min_line: len(ftext) - 1]` will never get the last line.

jeanconn replied:
And I think I need to add 1 to the context_range[1] to get a full ten lines after the line that includes the warning.

taldcroft commented on get_checks.py:
```diff
@@ -0,0 +1,189 @@
+    return {'filename': fname,
+            'line_number': idx + 1,
+            'text': literal_str("\n".join(warn_text)),
+            'context': context}
```
You can do the same thing for `context` that you do for `text` in the line above.

taldcroft commented on get_checks.py:
```diff
@@ -0,0 +1,189 @@
+        for idx, line in enumerate(text):
+            if re.match("^\s*#", line):
+                continue
+            if (re.search('[^\$]warning\s*\(', line)
```
DRY and use functional forms.

```
re_warns = ['[^\$]warning\s*\(', '...', '...', '...', '...']
if any(re.search(re_warn, line) for re_warn in re_warns):
```

taldcroft commented on get_checks.py:
```diff
@@ -0,0 +1,189 @@
+                    print("Conflict on id {}".format(c['id']))
+
+
+def confirm_no_lost_checks(checks, previous_checks):
```
Do we really need this?  It will certainly show up in the git diff.

jeanconn replied:
"Need" is relative in all of this code.  I'd like to keep the method as I'd prefer that my edits to the yaml were mostly single key edits instead of finding and copying over whole blocks.  That's what this method does now.  It could use a rename though, as it isn't confirming lost checks, it is copying over manual and missing ones.

jeanconn replied:
Though I suppose I take your point that these changed could be managed via the git diff.   I suppose I was somewhat solving the problem we don't have (keeping this process robust to future format changes independent of the git management of the yaml file).  I'll think upon it some more.

jeanconn commented on get_checks.py:
```diff
@@ -0,0 +1,189 @@
+                    c['id'] = pc['id']
+                    copy_manual_fields(c, pc)
+                else:
+                    print("Conflict on id {}".format(c['id']))
```
FYI, I was amused to see that we actually do have a "conflict" here as we presently have one warning that is called twice from the code with the same text (in two different contexts).

jeanconn committed:
Remove unneeded operation to find max of context range

jeanconn committed:
Simplify code around context lines

jeanconn committed:
Fix docstring typo

jeanconn committed:
Call lines of text 'lines' where appropriate

jeanconn committed:
Add 1 to max_line to actually get that line too

taldcroft commented:
Looking at the code now it seems like the integer key is more trouble than it's worth.  How about this:
- Use an 8-character hash based on the warning/error text as the primary `key` for an entry.
- Keep the YAML data structure as a list of dict.  (It's possible to use an OrderedDict but then you need to fuss around with the YAML loader and dumper like in http://stackoverflow.com/questions/5121931/in-python-how-can-you-load-yaml-mappings-as-ordereddicts).  
- You can get around this by making an OrderedDict `checks` after reading that maps `key` to the entry, and then at the end serialize `checks.values()` with YAML.
- Treat the data structure as any persistent database and do the updates in-place.
- For manual entries use a convention that the `key` starts with `m`, e.g. `m001`.  Define a function `is_manual` that distinguishes between a manual `key` and an auto-generated one.

With these changes you should be able to remove the four functions from lines 121 to 172.

The only new thing you would need is a function to remove entries that are no longer found.  In processing just accumulate a set of `keys` that have been seen and at the end remove any database entries that are not manual or have a key that is not in `keys` (issuing an informational message at that time).

taldcroft commented on make_html.py:
```diff
@@ -0,0 +1,57 @@
+
+checks = yaml.load(open(opt.infile).read())
+
+displaycols = ['id', 'link', 'filename', 'line_number',
```
`displaycols` can be (in order):
- `title` which is linked to the full entry
- `file` which is in the form `filename : line_number` and is linked to the `github_link`
- `type`
- `severity`
- `aca_checklist`

If we ever have meaningful content in `hopper` that might be added as a column.

taldcroft replied:
`ignore`, `missing` and `byhand` will no longer exist.

jeanconn replied:
Regarding "missing", if we remove a warning from the code, do you want the "permalink" to it to still work?  Should I represent this by keeping the warning but setting the filename to null or some such?

taldcroft replied:
The permalink should fail.  When hopper gets further along we can have this code find and warn about broken links.

taldcroft commented:
Code indentation is lost, e.g. http://cxc.cfa.harvard.edu/mta/ASPECT/review_aca_checks/first_draft/00012_check.html

jeanconn commented:
Ah.  I should replace the tabs with spaces before I strip.

jeanconn commented:
The biggest problem with "Use an 8-character hash based on the warning/error text as the primary key for an entry." is that we actually have a warning used twice.

jeanconn commented:
Also, I thought you wanted the id to be persistent even if the error message changes a bit.  Would you want these hashes to be initialized by the warning/error text, but then we'd let them drift (hash no longer would match new hash of warning text) by not updating that part of the yaml file using the git versioning?  If they can drift, I suppose I could just "make something up" for the duplicate warning.  Though at that point, I don't see the advantages of a hash over integer ids.

taldcroft commented:
A persistent id was the goal, but I'm not sure how to actually do that. Your existing code must have the same problem because you need some way to uniquely map each new check to a previous check in order to copy the old fields.  Maybe you've solved that?

In practice the text of existing warnings are quite stable.  In the unlikely event that this occurs we can modify the existing `checks.yaml` file to put in the hash of the new message so that it gets persisted properly.  E.g.

```
./update_checks.py  # this is doing an in-place update by default, so "update" seems good
git diff
# oops, unexpected change in one entry because of a different warning message
git checkout checks.yaml
emacs checks.yaml  # fix the key
./update_checks.py
./make_html.py --install  # makes new web pages locally and installs to $SKA/web/..
```

taldcroft commented:
`checks.yaml` should have the SHA tag of the starcheck repo and (ideally) the text starcheck version.  These would then go into the output HTML.

taldcroft commented:
We can finesse the duplicate warning text issue by either:
- Introducing a non-code-changing update into starcheck that makes them unique.
- Adding some lines of context to the warning text for making the key.

jeanconn commented:
I'd be on-board with a code-changing update for starcheck in this case.
The "second" warning would be improved by being specific about its context.

On Wed, Jun 29, 2016 at 11:15 AM, Tom Aldcroft notifications@github.com
wrote:

> We can finesse the duplicate warning text issue by either:
> - Introducing a non-code-changing update into starcheck that makes
>   them unique.
> - Adding some lines of context to the warning text for making the key.
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/sot/review_aca_checks/pull/1#issuecomment-229388634,
> or mute the thread
> https://github.com/notifications/unsubscribe/AAwT1Bt57IGoqd0c8dTfnLEDUZ1DmKkCks5qQov9gaJpZM4I-Gdq
> .

taldcroft commented:
> I'd be on-board with a code-changing update for starcheck in this case.
> The "second" warning would be improved by being specific about its context.

Even better!

jeanconn committed:
Only strip newlines from context

jeanconn commented:
Regarding: "A persistent id was the goal, but I'm not sure how to actually do that. Your existing code must have the same problem because you need some way to uniquely map each new check to a previous check in order to copy the old fields. Maybe you've solved that?"

The setup (to copy the ids to the new warnings on text matches only if the new warning didn't already have an id,) sort of solved this problem for the duplicateand didn't make it hard to proceed in the case of warning updates.   If a warning had new text it would be assigned a new id and the old one would be defined as missing.  We could then update by either changing the text in the old warning and removing the missing key, or we could remove the missing/old warning and update the id in the new text.  

But I agree that warning updates seem a weird priority.  If we don't care if the sha id is guaranteed to match the sha you'd get from the warning text, I think we retain flexibility and keep this code simple.

taldcroft commented:
Maybe we can do a hybrid solution:
- Use your incrementing `id` for the permalink.
- Use `text` as the unique key.  No need to generate a hash representation in this case.
- Temporarily define a `missing` field that will let us manually fix the `id` of the new entry if the old one went missing because of a change to the `text`.  The final committed file will have all missing entries deleted.

This does allow us to maintain a persistent permalink with nice URLs.  We would still have manually generated `id` in some cases.

taldcroft commented:
Ooops, that does not automatically propagate the other fields in the changed-`text` case.  Never mind.

taldcroft commented:
> If we don't care if the sha id is guaranteed to match the sha you'd get from the warning text, I think we retain flexibility and keep this code simple.

We do care that the SHA `id` matches the SHA from `text`.  The only time it would not match is during the temporary workaround process I described.

But now maybe I see a way around the automatic propagation issue by applying my workaround process, but using `text` as the primary key instead of the SHA of `text`.  Then `id` is just another field that gets generated along the way for any new entry that doesn't already exist.

In case it isn't obvious, the whole point of all this is to allow the entire update to a check to be essentially:

```
new_check = get_new_check(...)  # this creates only fields that have non-trivial data
text = new_check['text']
if text not in checks:
    new_check['id'] = get_next_id(...)
    checks[text] = new_check
else:
    checks[text].update(new_check)
```

jeanconn commented:
> But now maybe I see a way around the automatic propagation issue by applying my workaround process, but using text as the primary key instead of the SHA of text. Then id is just another field that gets generated along the way for any new entry that doesn't already exist.

I don't understand how this is any different from using the sha of
text (I mean I don't want to use the sha of text, but I've got to be
missing something if I'm not understanding your point at all).

taldcroft commented:
The point is using `text` as the key and `id` as the permalink.  In this case there is no need to generate a SHA tag, so it can be removed (one tiny step toward perfection).

jeanconn commented:
Though since the sha didn't get in there yet, I'm not sure if we get
points for removing it.​  :-)

Do you still want to use the format of the id (such as m001) to separate
"manual" entries from regular entries?

jeanconn committed:
Remove 'missing' field idea

Replace this idea with use of id beginning with 'm'

jeanconn commented:
I suppose "still" is a stretch in "Do you still want to use the format
of the id (such as m001) to separate "manual" entries from regular
entries?" because your original idea was to use that as text in the
sha key and now I'm talking about the id, but still...

jeanconn commented:
Also, if you get a chance for more explaining... I'm not clear what OrderedDict buys us.

taldcroft commented:
OrderedDict just keeps the entries in order.  This is assuming that one reads the **list** of entries from YAML and immediately constructs an OrderedDict from that list (where the dict is named `checks`).  Then all processing uses the OrderedDict `checks`, and finally YAML serializes `checks.values()` back out.  In that way there are no diffs introduced by random `dict` ordering.

jeanconn commented:
Gotcha.  I figured we'd put them out in "id" order either way and once those ids were filled in, the observations wouldn't shift anyway.  But I realize at this point that we are talking about like 5 different possible implementations of this code at once.  I got a short ways into totally redo-ing the data structures before getting your "perfect is the enemy of good" correspondence, so I'll probably table that reorganization for now.

jeanconn committed:
Append instead of extend to propagate previous checks

jeanconn committed:
Add code to get starcheck repo tag(s) if available

jeanconn committed:
Change the output structure to be a dict

Change the output structure to be a dictionary that contains the checks array
(as the value under the 'checks' key).

jeanconn committed:
Use functional form for warn matches

jeanconn committed:
Remove 'hopper' and 'ignore' from keys for now

jeanconn committed:
Simplify display cols and put individual reports out to one page

jeanconn committed:
Add starcheck version to top page

jeanconn committed:
Change the ids to strings

jeanconn committed:
Sort the checks by id before writing out

jeanconn committed:
Add has_match check on copying id and fields

This should prevent writing over manual data in the case of duplicate
text.

jeanconn committed:
Recast fields as literal strings if they will be written out again

jeanconn committed:
Remove 'text' and 'notes' from summary

jeanconn committed:
Fix one lingering 'str_id'

taldcroft commented:
Can you make the detail table be a 2-column vertical table with field name as the first column and value as the second.  That will be more readable and scale to when there is a real `note` or something.  Also there is no value in the `text` field in the detail table (it is unformatted code and just repeats what is shown in the code listing).

jeanconn committed:
Make details table a vertical table

jeanconn committed:
Remove 'text' from detail display

jeanconn commented:
Sure.  Up at [third_draft](http://cxc.cfa.harvard.edu/mta/ASPECT/review_aca_checks/third_draft/index.html)

I thought the dictionary list I had originally was confusing when there were missing values, but I also preferred the vertical overall display of that content

jeanconn committed:
Add code to save and print orvdot

jeanconn committed:
Remove 'missing' from printed items

jeanconn committed:
Add title to each entry's details

jeanconn committed:
Remove link from title to details (leave id link)

jeanconn committed:
Add code to allow missing context and text in manual entries

jeanconn committed:
Add code to warn when deleting a field

jeanconn committed:
Changes 'notes' to 'note' in yaml

jeanconn committed:
Save all key/values unless in given list

jeanconn committed:
Identify the has_match with idx and break when one found

jeanconn committed:
Disable default_flow_style

This seems to just effect the output of "manual" entries and helps to
put them out in the more-readable block style.

jeanconn committed:
Add title

jeanconn committed:
Add key and title to index page

jeanconn committed:
Add stylesheet

jeanconn committed:
Shorten filename display to basename
